storage/fanin/fanin.go

```diff
@@ -0,0 +1,179 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fanin
+
+import (
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/storage/local"
+	"github.com/prometheus/prometheus/storage/metric"
+	"github.com/prometheus/prometheus/storage/remote"
+)
+
+// Queryable is a local.Queryable that reads from local and remote storage.
+type Queryable struct {
+	Local  promql.Queryable
+	Remote *remote.Reader
+}
+
+// Querier implements local.Queryable.
+func (q Queryable) Querier() (local.Querier, error) {
+	localQuerier, err := q.Local.Querier()
+	if err != nil {
+		return nil, err
+	}
+
+	fq := querier{
+		local:   localQuerier,
+		remotes: q.Remote.Queriers(),
+	}
+	return fq, nil
+}
+
+type querier struct {
+	local   local.Querier
+	remotes []local.Querier
+}
+
+func (q querier) QueryRange(ctx context.Context, from, through model.Time, matchers ...*metric.LabelMatcher) ([]local.SeriesIterator, error) {
+	return q.query(func(q local.Querier) ([]local.SeriesIterator, error) {
+		return q.QueryRange(ctx, from, through, matchers...)
+	})
+}
+
+func (q querier) QueryInstant(ctx context.Context, ts model.Time, stalenessDelta time.Duration, matchers ...*metric.LabelMatcher) ([]local.SeriesIterator, error) {
+	return q.query(func(q local.Querier) ([]local.SeriesIterator, error) {
+		return q.QueryInstant(ctx, ts, stalenessDelta, matchers...)
+	})
+}
+
+func (q querier) query(qFn func(q local.Querier) ([]local.SeriesIterator, error)) ([]local.SeriesIterator, error) {
+	if len(q.remotes) == 0 {
+		// Skip potentially expensive merge logic if there are no remote queriers.
+		return qFn(q.local)
+	}
+
+	fpToIt := map[model.Fingerprint]iterator{}
+	for _, q := range append([]local.Querier{q.local}, q.remotes...) {
+		its, err := qFn(q)
+		if err != nil {
+			return nil, err
+		}
+		mergeIterators(fpToIt, its)
+	}
+
+	its := make([]local.SeriesIterator, 0, len(fpToIt))
+	for _, it := range fpToIt {
+		its = append(its, it)
+	}
+	return its, nil
+}
+
+func (q querier) MetricsForLabelMatchers(ctx context.Context, from, through model.Time, matcherSets ...metric.LabelMatchers) ([]metric.Metric, error) {
+	// TODO: implement querying metrics from remote storage.
+	return q.local.MetricsForLabelMatchers(ctx, from, through, matcherSets...)
+}
+
+func (q querier) LastSampleForLabelMatchers(ctx context.Context, cutoff model.Time, matcherSets ...metric.LabelMatchers) (model.Vector, error) {
+	// TODO: implement querying last samples from remote storage.
```

Comment:
That's not needed I think given it's just for federation.
For the new storage I brought up yesterday (https://github.com/fabxc/tsdb/issues/9) that it possibly shouldn't be a part of the main storage interface.